app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayMiddleware.kt

```diff
@@ -0,0 +1,72 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.tabstray
+
+import androidx.annotation.VisibleForTesting
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import org.mozilla.fenix.components.metrics.Event
+import org.mozilla.fenix.components.metrics.MetricController
+
+/**
+ * [Middleware] that reacts to various [TabsTrayAction]s.
+ *
+ * @property metrics reference to the configured [MetricController] to record general page load events.
+ */
+class TabsTrayMiddleware(
+    private val metrics: MetricController
+) : Middleware<TabsTrayState, TabsTrayAction> {
+
+    private var shouldReportInactiveTabMetrics: Boolean = true
+    private var shouldReportSearchGroupMetrics: Boolean = true
+
+    override fun invoke(
+        context: MiddlewareContext<TabsTrayState, TabsTrayAction>,
+        next: (TabsTrayAction) -> Unit,
+        action: TabsTrayAction
+    ) {
+        next(action)
+
+        when (action) {
+            is TabsTrayAction.UpdateInactiveTabs -> {
+                if (shouldReportInactiveTabMetrics) {
+                    shouldReportInactiveTabMetrics = false
+                    metrics.track(Event.InactiveTabsCountUpdate(action.tabs.size))
+                    metrics.track(Event.TabsTrayHasInactiveTabs(action.tabs.size))
+                }
+            }
+            is TabsTrayAction.UpdateSearchGroupTabs -> {
+                if (shouldReportSearchGroupMetrics) {
+                    shouldReportSearchGroupMetrics = false
+
+                    metrics.track(Event.SearchTermGroupCount(action.groups.size))
+
+                    if (action.groups.isNotEmpty()) {
+                        val tabsPerGroup = action.groups.map { it.tabs.size }
+                        val averageTabsPerGroup = tabsPerGroup.average()
+                        metrics.track(Event.AverageTabsPerSearchTermGroup(averageTabsPerGroup))
+
+                        val tabGroupSizeMapping = tabsPerGroup.map { generateTabGroupSizeMappedValue(it) }
+                        metrics.track(Event.SearchTermGroupSizeDistribution(tabGroupSizeMapping))
+                    }
+                }
+            }
+            else -> {}
```

Comment:
nit: we don't need this else despite what Android Studio says. We don't do it for any other middleware.

Reply:
kk. good to know for the future; will remove. Thanks again for the review!